File: backend/src/communication/MessageData.kt
package com.colegawin.remotekontrol.backend.communication

import com.colegawin.remotekontrol.backend.server.ServerState

open class MessageData
data class TransitionData(val state: ServerState, val payload: String = "") : MessageData()
data class InstructionData(val drive: Int, val rotation: Int) : MessageData()
data class UpdateData(val drive: Int, val rotation: Int) : MessageData()
data class ErrorData(val reason: ErrorType, val message: String) : MessageData()

File: client/lib/models/messages/message_data.dart
import 'package:client/ctrls/ctrls.dart';

abstract class MessageData {
  Map<String, dynamic> toJson();
}

class InstructionData extends MessageData {
  final int drive;
  final int rotation;

  InstructionData(this.drive, this.rotation);

  @override
  Map<String, dynamic> toJson() => {'drive': drive, 'rotation': rotation};
}

class UpdateData extends InstructionData {
  UpdateData(int drive, int rotation) : super(drive, rotation);
}

class TransitionData extends MessageData {
  final ServerState state;

  TransitionData(this.state);

  @override
  Map<String, dynamic> toJson() => {'state': state.toString().substring(12)};
}
